pkg/sources/github/feedlet.go

```diff
@@ -0,0 +1,419 @@
+				continue
+			}
+			status := service.Status
+			if status.Domain != "" {
+				w.Stop()
+				log.Printf("got domain: %s", event.Object)
+				return status.Domain, nil
+			}
+
+		}
+	}
+}
+
+func receiveAdapterName(resource string) string {
+	// TODO(n3wscott): this needs more UUID on the end of it?
+	// TODO(n3wscott): Currently this needs to be deterministic so StopFeed can find the receive adapter. If the receive adapter name were added to the feed context, then this could be a uuid.
+	serviceName := fmt.Sprintf("%s-%s-%s", "github", resource, postfixReceiveAdapter)
+	serviceName = strings.Replace(serviceName, "/", "-", -1)
+	serviceName = strings.Replace(serviceName, ".", "-", -1)
+	serviceName = strings.ToLower(serviceName)
+	return serviceName
+}
+
+func (t *GithubEventSource) createReceiveAdapter(trigger sources.EventTrigger, target string) (*v1alpha1.Service, error) {
+	sc := t.servingclientset.ServingV1alpha1().Services(t.feedNamespace)
+
+	serviceName := receiveAdapterName(trigger.Resource)
+
+	// First, check if service exists already.
+	if _, err := sc.Get(serviceName, metav1.GetOptions{}); err != nil {
+		if !apierrs.IsNotFound(err) {
+			return nil, fmt.Errorf("service.Get for %q failed: %v", serviceName, err)
+		}
+		log.Printf("service %q doesn't exist, creating", serviceName)
+	} else {
+		log.Printf("found existing service %q", serviceName)
+		// Don't try again.
+		return nil, nil
+	}
+
+	secretName, err := stringFrom(trigger.Parameters, secretNameKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secret name from trigger parameters: %v", err)
+	}
+
+	secretKey, err := stringFrom(trigger.Parameters, secretKeyKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secret key from trigger parameters: %v", err)
+	}
+
+	service := resources.MakeService(t.feedNamespace, serviceName, t.image, t.feedServiceAccountName, target, secretName, secretKey)
+	service, err = sc.Create(service)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create service: %s", err)
+	}
+
+	return service, nil
+}
+
+func (t *GithubEventSource) deleteReceiveAdapter(namespace string, serviceName string) error {
+	sc := t.servingclientset.ServingV1alpha1().Services(t.feedNamespace)
+
+	// First, check if deployment exists already.
+	if _, err := sc.Get(serviceName, metav1.GetOptions{}); err != nil {
+		if !apierrs.IsNotFound(err) {
+			return fmt.Errorf("services.Get for %q failed: %v", serviceName, err)
+		}
+		log.Printf("service %q already deleted", serviceName)
+		// Don't try again.
+		return nil
+	}
+
+	return sc.Delete(serviceName, &metav1.DeleteOptions{})
+}
+
+func (t *GithubEventSource) deleteWebhook(trigger sources.EventTrigger, feedContext sources.FeedContext) error {
+	serviceName := receiveAdapterName(trigger.Resource)
+	t.deleteReceiveAdapter(t.feedNamespace, serviceName)
+
+	owner, repo, err := parseOwnerRepoFrom(trigger.Resource)
```

Comment:
I am blocking the creation of the webhook on the successful creation of the Receive Adapter. I was trying to be clean about it but it is never fun... I think this could get refactored to be a bit more clear what is happening.